cmd/tools/doctor/doctor.go

```diff
@@ -53,7 +69,110 @@ func checkAll(path string, contents []byte) {
 		os.Exit(1)
 		return
 	}
-	os.Exit(0)
+
+	anyFailed := false
+	anyFailed = !checkUniquePromPorts(*harvestConfig).isValid || anyFailed
+	anyFailed = !checkExporterTypes(*harvestConfig).isValid || anyFailed
+
+	if anyFailed {
+		os.Exit(1)
+	} else {
+		os.Exit(0)
+	}
+}
+
+func detectConsole() {
+	withColor = false
+	switch opts.Color {
+	case "never":
+		withColor = false
+	case "always":
+		withColor = true
+	default:
+		if term.IsTerminal(int(os.Stdout.Fd())) {
+			withColor = true
+		}
+	}
+}
+
+// checkExporterTypes validates that all exporters are of valid types
+func checkExporterTypes(config conf.HarvestConfig) validation {
+	if config.Exporters == nil {
+		return validation{}
+	}
+	invalidTypes := make(map[string]string)
+	for name, exporter := range *config.Exporters {
+		if exporter.Type == nil {
+			continue
+		}
+		switch *exporter.Type {
+		case "Prometheus", "InfluxDB":
+			break
+		default:
+			invalidTypes[name] = *exporter.Type
+		}
+	}
+
+	valid := validation{isValid: true}
+
+	if len(invalidTypes) > 0 {
+		valid.isValid = false
+		fmt.Printf("%s Unknown Exporter types found\n", colorize("Error:", color.Red))
+		fmt.Println("These are probably misspellings or the wrong case.")
+		fmt.Println("Exporter types must start with a capital letter.")
+		fmt.Println("The following exporters are unknown:")
+		for name, eType := range invalidTypes {
+			valid.invalid = append(valid.invalid, eType)
+			fmt.Printf("  exporter named: [%s] has unknown type: [%s]\n", colorize(name, color.Red), colorize(eType, color.Yellow))
+		}
+		fmt.Println()
+	}
+	return valid
+}
+
+// checkUniquePromPorts checks that all Prometheus exporters
+// that specify a port, do so uniquely
+func checkUniquePromPorts(config conf.HarvestConfig) validation {
+	if config.Exporters == nil {
+		return validation{}
+	}
+	// Add all exporters that have a port to a
+	// map of portNum -> list of names
+	seen := make(map[int][]string)
+	for name, exporter := range *config.Exporters {
+		if exporter.Port == nil || exporter.Type == nil || *exporter.Type != "Prometheus" {
+			continue
+		}
+		previous := seen[*exporter.Port]
+		previous = append(previous, name)
```

Comment:
previous may be nil. It will work but why to add nil to slice.

Reply:
yes, previous will be nil when the map is empty
but `append` turns that nil slice into a new one
that's idiomatic Go and cleaner than if previous = nil {create new slice}